src/atlas/array/ArrayViewVariant.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * (C) Crown Copyright 2024 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <variant>
+
+#include "atlas/array.h"
+
+namespace atlas {
+namespace array {
+
+namespace detail {
+
+using namespace array;
+
+// Container struct for a list of types.
+template <typename... Ts>
+struct Types {};
+
+// Container struct for a list of integers.
+template <int... Is>
+struct Ints {};
+
+template <typename Values, typename Ranks, typename... ArrayViews>
+struct VariantHelper;
+
+// Recursively construct ArrayView std::variant from types Ts and ranks Is.
+template <typename T, typename... Ts, int... Is, typename... ArrayViews>
+struct VariantHelper<Types<T, Ts...>, Ints<Is...>, ArrayViews...> {
+  using type = typename VariantHelper<Types<Ts...>, Ints<Is...>, ArrayViews...,
+                                      ArrayView<const T, Is>...,
+                                      ArrayView<T, Is>...>::type;
+};
+
+// End recursion.
+template <int... Is, typename... ArrayViews>
+struct VariantHelper<Types<>, Ints<Is...>, ArrayViews...> {
+  using type = std::variant<ArrayViews...>;
+};
+
+template <typename Values, typename Ranks>
+using Variant = typename VariantHelper<Values, Ranks>::type;
+
+}  // namespace detail
+
+/// @brief Supported ArrayView value types.
+using Values = detail::Types<float, double, int, long, unsigned long>;
+
+/// @brief Supported ArrayView ranks.
+using Ranks = detail::Ints<1, 2, 3, 4, 5, 6, 7, 8, 9>;
+
+/// @brief Variant containing all supported ArrayView alternatives.
+using ArrayViewVariant = detail::Variant<Values, Ranks>;
+
```

Comment:
In case you're interested, `array::ArrayViewVariant` currently expands into:

```
std::variant<atlas::array::ArrayView<float const, 1>,
             atlas::array::ArrayView<float const, 2>,
             atlas::array::ArrayView<float const, 3>,
             atlas::array::ArrayView<float const, 4>,
             atlas::array::ArrayView<float const, 5>,
             atlas::array::ArrayView<float const, 6>,
             atlas::array::ArrayView<float const, 7>,
             atlas::array::ArrayView<float const, 8>,
             atlas::array::ArrayView<float const, 9>,
             atlas::array::ArrayView<float, 1>,
             atlas::array::ArrayView<float, 2>,
             atlas::array::ArrayView<float, 3>,
             atlas::array::ArrayView<float, 4>,
             atlas::array::ArrayView<float, 5>,
             atlas::array::ArrayView<float, 6>,
             atlas::array::ArrayView<float, 7>,
             atlas::array::ArrayView<float, 8>,
             atlas::array::ArrayView<float, 9>,
             atlas::array::ArrayView<double const, 1>,
             atlas::array::ArrayView<double const, 2>,
             atlas::array::ArrayView<double const, 3>,
             atlas::array::ArrayView<double const, 4>,
             atlas::array::ArrayView<double const, 5>,
             atlas::array::ArrayView<double const, 6>,
             atlas::array::ArrayView<double const, 7>,
             atlas::array::ArrayView<double const, 8>,
             atlas::array::ArrayView<double const, 9>,
             atlas::array::ArrayView<double, 1>,
             atlas::array::ArrayView<double, 2>,
             atlas::array::ArrayView<double, 3>,
             atlas::array::ArrayView<double, 4>,
             atlas::array::ArrayView<double, 5>,
             atlas::array::ArrayView<double, 6>,
             atlas::array::ArrayView<double, 7>,
             atlas::array::ArrayView<double, 8>,
             atlas::array::ArrayView<double, 9>,
             atlas::array::ArrayView<int const, 1>,
             atlas::array::ArrayView<int const, 2>,
             atlas::array::ArrayView<int const, 3>,
             atlas::array::ArrayView<int const, 4>,
             atlas::array::ArrayView<int const, 5>,
             atlas::array::ArrayView<int const, 6>,
             atlas::array::ArrayView<int const, 7>,
             atlas::array::ArrayView<int const, 8>,
             atlas::array::ArrayView<int const, 9>,
             atlas::array::ArrayView<int, 1>,
             atlas::array::ArrayView<int, 2>,
             atlas::array::ArrayView<int, 3>,
             atlas::array::ArrayView<int, 4>,
             atlas::array::ArrayView<int, 5>,
             atlas::array::ArrayView<int, 6>,
             atlas::array::ArrayView<int, 7>,
             atlas::array::ArrayView<int, 8>,
             atlas::array::ArrayView<int, 9>,
             atlas::array::ArrayView<long const, 1>,
             atlas::array::ArrayView<long const, 2>,
             atlas::array::ArrayView<long const, 3>,
             atlas::array::ArrayView<long const, 4>,
             atlas::array::ArrayView<long const, 5>,
             atlas::array::ArrayView<long const, 6>,
             atlas::array::ArrayView<long const, 7>,
             atlas::array::ArrayView<long const, 8>,
             atlas::array::ArrayView<long const, 9>,
             atlas::array::ArrayView<long, 1>,
             atlas::array::ArrayView<long, 2>,
             atlas::array::ArrayView<long, 3>,
             atlas::array::ArrayView<long, 4>,
             atlas::array::ArrayView<long, 5>,
             atlas::array::ArrayView<long, 6>,
             atlas::array::ArrayView<long, 7>,
             atlas::array::ArrayView<long, 8>,
             atlas::array::ArrayView<long, 9>,
             atlas::array::ArrayView<unsigned long const, 1>,
             atlas::array::ArrayView<unsigned long const, 2>,
             atlas::array::ArrayView<unsigned long const, 3>,
             atlas::array::ArrayView<unsigned long const, 4>,
             atlas::array::ArrayView<unsigned long const, 5>,
             atlas::array::ArrayView<unsigned long const, 6>,
             atlas::array::ArrayView<unsigned long const, 7>,
             atlas::array::ArrayView<unsigned long const, 8>,
             atlas::array::ArrayView<unsigned long const, 9>,
             atlas::array::ArrayView<unsigned long, 1>,
             atlas::array::ArrayView<unsigned long, 2>,
             atlas::array::ArrayView<unsigned long, 3>,
             atlas::array::ArrayView<unsigned long, 4>,
             atlas::array::ArrayView<unsigned long, 5>,
             atlas::array::ArrayView<unsigned long, 6>,
             atlas::array::ArrayView<unsigned long, 7>,
             atlas::array::ArrayView<unsigned long, 8>,
             atlas::array::ArrayView<unsigned long, 9>>;
```

Reply:
This has now been split into const and non-const variants!